Sources/_StringProcessing/Regex/Match.swift

```diff
@@ -37,33 +41,48 @@ extension Regex.Match {
       )
       
       let output = AnyRegexOutput(
-        input: anyRegexOutput.input,
+        input: input,
         elements: [wholeMatchCapture] + anyRegexOutput._elements
       )
       
       return output as! Output
     } else if Output.self == Substring.self {
       // FIXME: Plumb whole match (`.0`) through the matching engine.
-      return anyRegexOutput.input[range] as! Output
-    } else if anyRegexOutput.isEmpty, value != nil {
+      return input[range] as! Output
+    } else if anyRegexOutput.isEmpty, let value {
       // FIXME: This is a workaround for whole-match values not
       // being modeled as part of captures. We might want to
       // switch to a model where results are alongside captures
-      return value! as! Output
+      return value as! Output
     } else {
       guard value == nil else {
         fatalError("FIXME: what would this mean?")
       }
       let typeErasedMatch = anyRegexOutput.existentialOutput(
-        from: anyRegexOutput.input[range]
+        from: input[range]
       )
       return typeErasedMatch as! Output
     }
   }
 
+  var wholeMatchType: Any.Type {
+    value.map { type(of: $0) } ?? Substring.self
```

Comment:
What if it's optional? Also, are we tracking nested optionals correctly here (I.e. do they get turned into values or are they counted)?

Reply:
When there's no dedicated value for whole match, it can only be a `Substring`.